tracer/src/Datadog.Trace/AppSec/ApiSecurity.cs

```diff
@@ -6,44 +6,108 @@
+    private readonly int _maxRoutesSize;
     private readonly bool _enabled;
-    private readonly OverheadController _overheadController;
+    private readonly TimeSpan _minTimeBetweenReprocessTimeSpan;
+    private readonly OrderedDictionary _processedRoutes = new();
+    private readonly Dictionary<string, bool> _apiSecurityAddress = new() { { "extract-schema", true } };
 
-    public ApiSecurity(SecuritySettings securitySettings)
+    public ApiSecurity(SecuritySettings securitySettings, int maxRouteSize = 4096)
     {
-        _overheadController = new(securitySettings.ApiSecurityMaxConcurrentRequests, (int)(securitySettings.ApiSecuritySampling * 100));
         // todo: later, will be enabled by default, depending on if Security is enabled
         _enabled = securitySettings.ApiSecurityEnabled;
+        _minTimeBetweenReprocessTimeSpan = TimeSpan.FromSeconds(securitySettings.ApiSecuritySampleDelay);
+        _maxRoutesSize = maxRouteSize;
     }
 
-    public bool TryTellWafToAnalyzeSchema(IDictionary<string, object> args)
+    public bool ShouldAnalyzeSchema(bool lastWafCall, Span localRootSpan, IDictionary<string, object> args, string? statusCode, IDictionary<string, object>? routeValues)
     {
-        if (_enabled && _overheadController.AcquireRequest())
+        try
         {
-            args.Add(AddressesConstants.WafContextProcessor, new Dictionary<string, bool> { { "extract-schema", true } });
-            return true;
-        }
+            var samplingPriority = localRootSpan.Context.TraceContext.SamplingPriority;
+            if (_enabled && lastWafCall && samplingPriority is SamplingPriorityValues.AutoKeep or SamplingPriorityValues.UserKeep)
+            {
+                var httpRouteTag = localRootSpan.GetTag(Tags.AspNetCoreEndpoint) ?? localRootSpan.GetTag(Tags.HttpRoute);
+                var httpMethod = localRootSpan.GetTag(Tags.HttpMethod);
+                statusCode ??= localRootSpan.GetTag(Tags.HttpStatusCode);
+                if (httpRouteTag == null || httpMethod == null || statusCode == null)
+                {
+                    Log.Debug("Unsupported groupkey for api security {Route}, {Method}, {Status}", httpRouteTag, httpMethod, statusCode);
+                    return false;
+                }
 
-        return false;
-    }
+                if (routeValues != null)
+                {
+                    foreach (var routeValue in routeValues)
+                    {
+                        var routeKey = $"{{{routeValue.Key}}}";
+                        if (routeValue.Value is string && httpRouteTag.Contains(routeKey))
+                        {
+                            httpRouteTag = httpRouteTag.Replace(routeKey, routeValue.Value.ToString());
+                        }
+                    }
+                }
 
-    public void ReleaseRequest()
-    {
-        if (_enabled)
+                var hash = CombineHashes(httpRouteTag, httpMethod, statusCode);
```

Comment:
Maybe remove this helper method?

```suggestion
                var hash = HashCode.Combine(httpRouteTag, httpMethod, statusCode);
```

Reply:
I can't as it's used by the tests.. and I had a really odd problem I posted about, where
> I cant use the class HashCode as it's both in System.Runtime and Datadog.Trace (vendored) I get Type 'X' exists in both assemblies / Type 'HashCode' is ambiguous..
I then also tried linking `Datadog.Trace.SourceGenerators.Helpers.HashCode()` but it wouldn't work either..

Reply:
I see, thanks!